ellipsoid/v1/caesar/spheroid.hpp

```diff
@@ -0,0 +1,178 @@
+ * that approximates the size and shape of the Earth. In such applications, the
+ * semi-major axis length describes the equatorial radius and the semi-minor
+ * axis length describes the polar radius.
+ *
+ * Reference ellipsoids are commonly specified by their semi-major axis and
+ * flattening constant, \f$f\f$, defined as
+ *
+ * \f[
+ * f = \frac{a - b}{a}
+ * \f]
+ *
+ * The flattening is a measure of the ellipticity of the spheroid. If the
+ * flattening is between 0 and 1, the spheroid is oblate (\f$a > b\f$). If it's
+ * equal to 0, the spheroid is a perfect sphere (\f$a == b\f$). If it's less
+ * than zero, the spheroid is prolate (\f$a < b\f$).
+ *
+ * \see wgs84_ellipsoid
+ */
+class Spheroid {
+public:
+    /**
+     * Construct a new Spheroid object.
+     *
+     * \param[in] a semi-major axis length
+     * \param[in] f flattening constant
+     */
+    constexpr Spheroid(double a, double f) noexcept : a_(a), f_(f) {}
+
+    /**
+     * Return the semi-major axis length.
+     *
+     * In geodetic applications, this is the equatorial radius.
+     *
+     * \see Spheroid::a
+     */
+    [[nodiscard]] constexpr double
+    semimajor_axis() const noexcept
+    {
+        return a();
+    }
+
+    /** Same as Spheroid::semimajor_axis() */
+    [[nodiscard]] constexpr double
+    a() const noexcept
+    {
+        return a_;
+    }
+
+    /**
+     * Return the semi-minor axis length.
+     *
+     * In geodetic applications, this is the polar radius.
+     *
+     * \see Spheroid::b
+     */
+    [[nodiscard]] constexpr double
+    semiminor_axis() const noexcept
+    {
+        return b();
+    }
+
+    /** Same as Spheroid::semiminor_axis() */
+    [[nodiscard]] constexpr double
+    b() const noexcept
+    {
+        return a() * (1. - f());
+    }
+
+    /**
+     * Return the (first) flattening.
+     *
+     * \see Spheroid::f
+     */
+    [[nodiscard]] constexpr double
+    flattening() const noexcept
+    {
+        return f();
+    }
+
+    /** Same as Spheroid::flattening() */
+    [[nodiscard]] constexpr double
+    f() const noexcept
+    {
+        return f_;
+    }
+
+    /**
+     * Return the inverse flattening constant, \f$\frac{1}{f}\f$.
+     *
+     * \see Spheroid::flattening
+     */
+    [[nodiscard]] constexpr double
+    inverse_flattening() const
+    {
+        return 1. / f();
+    }
+
+    /**
+     * Return the third flattening.
+     *
+     * The third flattening, \f$n\f$, is defined as
+     *
+     * \f[
+     * n = \frac{a - b}{a + b}
+     * \f]
+     */
+    [[nodiscard]] constexpr double
+    third_flattening() const noexcept
+    {
+        return f() / (2. - f());
+    }
+
+    /**
+     * Return the eccentricity of the spheroid.
+     *
+     * The eccentricity, ]f$e\f$, is defined as
+     *
+     * \f[
+     * e = \sqrt{1 - \frac{b^2}{a^2}}
+     * \f]
+     */
+    [[nodiscard]] double
+    eccentricity() const
```

Comment:
```suggestion
    eccentricity() const noexcept
```

Reply:
Huh you're right, that's odd that it isn't noexcept. I assumed it was, after reading that even bad inputs will just return nan or rounded values. Does that mean that prolate spheroids have complex eccentricity then? Lol, weird

Reply:
> Does that mean that prolate spheroids have imaginary eccentricity then?

According to [this](https://en.wikipedia.org/wiki/Spheroid#Area), squared eccentricity is always positive, so I guess I was wrong. For a prolate spheroid, you flip the position of `a` and `b` in the equation. In other words, eccentricity is defined in terms of the semi-major and semi-minor axis lengths instead of `a` and `b`.

I think that also means that `flattening` should always be positive, in contradiction with what I wrote in the class docstring. Oops.